GSW/testUtility/testNS2.pyw

```diff
@@ -26,87 +26,148 @@ commandField = Entry(commandFrame, width=30, borderwidth=3, font=('Consolas',10)
+            updateLog('Connected to port \"' + portField.get() + '\".')
+            return    
+    updateLog('Port \"' + portField.get() + '\" not found.')
+    
+# save to file #
+def saveFile(): # saves monitor contents to a file
+    fileName = fileNameField.get() + '.txt'
+    filePath = 'saved files\\' + fileNameField.get()
+    if (os.path.exists(filePath)): # display warning if file already exists
+        warning = '\"' + fileName + '\" already exists. Do you wish to overwrite it?'
+        overwrite = tkinter.messagebox.askyesno(title='Warning', message= warning)
+        if (not overwrite):
+            return
+    # write to file
+    file = open(filePath,'w', newline = '')
+    file.write(monitor.get('1.0','end-1c'))
+    updateLog( 'Output saved to ' + fileName)
+
+# clear output #
+def clearOutput(): # clears the output monitor
+    confirm = tkinter.messagebox.askyesno(title='Warning', message= 'Are you sure you want to clear the output monitor???')
+    if confirm:
+        monitor.configure(state ='normal')
+        monitor.delete('1.0', 'end')
+        monitor.configure(state='disabled')
+        updateLog('Output monitor cleared.')
 
+# ==== buttons ====
+button_sendCommand = Button(commandFrame, text='Send', command=sendCommand)
+root.bind('<Return>', sendCommand) # bind the enter key to send command
+button_scanPorts = Button(serialFrame, text='Scan For Serial Ports', command=scanPorts)
+button_updatePort = Button(serialFrame, text='Open Serial Port', bg='#c9ffba', command=openSerialPort)
+button_saveFile = Button(fileFrame, text='Save to File', command=saveFile)
+button_clearOutput = Button(monitorFrame, text='Clear', command=clearOutput)
 
 # ==== draw everything ====
 # draw command frame
 commandFrame.grid(row=0, column=0, padx=5, pady=5, sticky=W+E)
 commandField.grid(row=0, column=0, padx=5, pady=3, sticky=W+E)
 button_sendCommand.grid(row=0, column=1, sticky=W+E)
 
-# log frame
+# draw log frame
 logFrame.grid(row=1, column=0, padx=5, pady=5, sticky=N+S+E+W)
 log.pack(expand=True, fill=BOTH)
 
 # draw monitor frame
-monitorFrame.grid(row=0, column=1, padx=5, pady=5, sticky=N+W+E+S, rowspan=3)
+monitorFrame.grid(row=0, column=1, padx=5, pady=5, sticky=N+W+E+S, rowspan=4)
 monitor.pack(expand=True, fill=BOTH)
+check_autoScroll.pack(side='left')
+button_clearOutput.pack(side='right')
 
 # draw serial frame
 serialFrame.grid(row=2, column=0, padx=5, pady=5, sticky=W+E)
 portField.grid(row=0, column=0, sticky=W+E)
 button_updatePort.grid(row=0, column=1, sticky=W+E)
 button_scanPorts.grid(row=1, column=0, sticky=W+E, columnspan=2)
 
+# draw file frame
+fileFrame.grid(row=3, column=0, padx=5, pady=5, sticky=W+E)
+fileNameField.grid(row=0, column=0, sticky=W+E)
+button_saveFile.grid(row=0, column=1, sticky=W+E)
 
 # ==== start application ====
-teensy = serial.Serial(port='COM3', baudrate=9600, timeout=10/1000)
+# configure serial connection
+portOpen = False
+teensy = serial.Serial()
+teensy.baudrate = 19200
```

Comment:
What prompted the baud rate change?

Reply:
I think think the only reason we had it at 9600 was because the Arduino examples had that. Turns out the Teensy can handle a much higher rate. So basically just because we can.